firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,491 @@
+        persisted = {
+            'update': {
+                'index': 0,
+                'staging_path': self.staging_directory},
+            # Create results object with information of the unmodified pre build
+            # TODO: Lets change to persisted.update.results once the issue is fixed:
+            # https://github.com/mozilla/mozmill-automation/issues/175
+            'updates': [
+                {
+                    'build_pre': self.build_info,
+                    'build_post': self.build_info,
+                    'fallback': fallback,
+                    'patch': {},
+                    'success': False}]}
+
+        # If requested modify the default update channel. It will be active
+        # after the next restart of the application
+        if persisted['update']['channel']:
+            # Backup the original content and the path of the channel-prefs.js file
+            persisted['update']['default_update_channel'] = {
+                'content': self.update_channel.file_contents,
+                'path': self.update_channel.file_path}
+            self.update_channel.default_channel = persisted['update']['channel']
+
+        # If requested modify the list of allowed MAR channels
+        if persisted['update']['allowed_mar_channels']:
+            # Backup the original content and the path of the update-settings.ini file
+            persisted['update']['default_mar_channels'] = {
+                'content': self.mar_channels.config_file_contents,
+                'path': self.mar_channels.config_file_path}
+
+    def get_update_url(self, force=False):
+        """Retrieve the AUS update URL the update snippet is retrieved from
+
+        :param force: Boolean flag to force an update check
+
+        :returns: The URL of the update snippet
+        """
+
+        url = self.prefs.get_pref(self.PREF_APP_UPDATE_URL, '')
+
+        if not url:
+            return None
+
+        # get the next two prefs from the default branch
+        dist = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION, True) or 'default'
+        dist_version = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION_VERSION, True) or 'default'
+
+        # Not all placeholders are getting replaced correctly by formatURL
+        url = url.replace('%PRODUCT%', self.app_info.name)
+        url = url.replace('%BUILD_ID%', self.app_info.appBuildID)
+        url = url.replace('%BUILD_TARGET%', self.app_info.OS + '_' + self.ABI)
+        url = url.replace('%OS_VERSION%', self.os_version)
+        url = url.replace('%CHANNEL%', self.update_channel.channel)
+        url = url.replace('%DISTRIBUTION%', dist)
+        url = url.replace('%DISTRIBUTION_VERSION%', dist_version)
+
+        url = self.marionette.execute_script("""
+          Cu.import("resource://gre/modules/Services.jsm");
+
+          return Services.urlFormatter.formatURL(arguments[0]);
+
+          """, script_args=[url])
```

Comment:
nit: indentation issue. Align it with `u` from url. Maybe drop the blank lines in the JS code.